Avoid duplicate reading for small parquet files

marin-ma opened the pull request:
For small parquet files, when loading footer, the whole file is already read to memory, no need to read file again when loading data.
Before this change:
![image](https://github.com/facebookincubator/velox/assets/52736607/964a1b01-5654-408f-af9c-c67f8d2b99f3)

After this change:
![image](https://github.com/facebookincubator/velox/assets/52736607/1bd1cb20-95af-42c6-afe6-af9cb3b525fd)

marin-ma commented:
cc: @JkSelf @zhli1142015

JkSelf commented:
@marin-ma Can you help to add some unit tests?

Yohahaha commented on velox/dwio/parquet/reader/ParquetReader.cpp:
```diff
@@ -47,12 +47,18 @@ ReaderBase::ReaderBase(
-      fileLength_ - readSize, readSize, dwio::common::LogType::FOOTER);
+  std::unique_ptr<dwio::common::SeekableInputStream> stream = nullptr;
+  if (preloadFile_) {
+    stream = input_->readFile(fileLength_, dwio::common::LogType::FOOTER);
```
Could we use `input_->read(0, fileLength_)` instead of create a new method `readFile`?

marin-ma replied:
`readFile` uses enqueue/load to populate underlying buffers, which is different from what `read` do.

majetideepak replied:
`dwio::common::LogType::FILE` since we are reading the entire file?

Yohahaha commented on velox/dwio/parquet/reader/ParquetReader.h:
```diff
@@ -109,16 +109,18 @@ class ReaderBase {
+
   // Map from row group index to pre-created loading BufferedInput.
-  std::unordered_map<uint32_t, std::unique_ptr<dwio::common::BufferedInput>>
+  std::unordered_map<uint32_t, std::shared_ptr<dwio::common::BufferedInput>>
```
Why this change is needed?

marin-ma replied:
For small files,`input_` buffers the whole file. Thus we can store the same BufferedInput ptr for each rowgroup.

marin-ma commented:
@Yuhta PTAL.

marin-ma commented:
@Yuhta Could you please take a look? Thanks!

marin-ma commented:
@mbasmanova Could you please take a look? We're waiting for this PR to get reviewed.

mbasmanova commented:
@marin-ma I'm out next week.

@majetideepak @Yuhta Deepak, Jimmy, would you help review this PR?

majetideepak commented:
@marin-ma I see that you exposed the ReaderBase interface for testing. Can we use File I/O metrics instead to check there are no duplicate reads? As you commented in the code, ReaderBase is internal only. I am adding a PR to move that from the header file to the CPP file as it is internal only.

marin-ma commented:
> @marin-ma I see that you exposed the ReaderBase interface for testing. Can we use File I/O metrics instead to check there are no duplicate reads? As you commented in the code, ReaderBase is internal only. I am adding a PR to move that from the header file to the CPP file as it is internal only.

Updated UT and removed the newly added internal-only APIs. PTAL. Thanks!

majetideepak commented on velox/dwio/parquet/reader/ParquetReader.cpp:
```diff
@@ -476,19 +482,30 @@ void ReaderBase::scheduleRowGroups(
-        reader.enqueueRowGroup(nextGroup, *newInput);
-        newInput->load(dwio::common::LogType::STRIPE);
-        inputs_[nextGroup] = std::move(newInput);
+      if (inputs_.count(nextGroup) == 0) {
```
Looks like you fixed a bug in the row group prefetching. We should fix this independently along with a test. Can you open an issue and undo this fix for this PR?

marin-ma replied:
Separated: https://github.com/facebookincubator/velox/pull/6121

majetideepak commented on velox/dwio/parquet/reader/ParquetReader.h:
```diff
@@ -109,16 +109,18 @@ class ReaderBase {
   // Copy of options. Must be owned by 'this'.
   const dwio::common::ReaderOptions options_;
-  std::unique_ptr<velox::dwio::common::BufferedInput> input_;
+  std::shared_ptr<velox::dwio::common::BufferedInput> input_;
```
Can this change be reverted to unique_ptr? Same below.

majetideepak reviewed (commented):
@marin-ma Some more comments.
It would be cleaner if we can avoid the branching for `preloadFile_`. Let me take a look.

majetideepak commented on velox/dwio/parquet/reader/ParquetReader.cpp:
```diff
@@ -47,12 +47,18 @@ ReaderBase::ReaderBase(
 void ReaderBase::loadFileMetaData() {
-  bool preloadFile_ = fileLength_ <= filePreloadThreshold_;
+  preloadFile_ = fileLength_ <= filePreloadThreshold_ ||
+      fileLength_ <= directorySizeGuess_;
```
also, why include `fileLength_ <= directorySizeGuess_`?

marin-ma replied:
@zhli1142015 @JkSelf Can you help to answer this question? Thanks!

marin-ma replied:
@majetideepak Because in the next line, `readSize` will be `fileLength_` if `fileLength_ <= directorySizeGuess_`. In this case the whole file is loaded.
```
  uint64_t readSize =
      preloadFile_ ? fileLength_ : std::min(fileLength_, directorySizeGuess_);
```
While the code looks weird. Optimized this part of logic in the latest commit.

majetideepak commented on velox/dwio/common/BufferedInput.h:
```diff
@@ -122,6 +122,14 @@ class BufferedInput {
     return std::make_unique<BufferedInput>(input_, pool_);
   }
 
+  std::unique_ptr<SeekableInputStream> readFile(
```
I don't think API fits the BufferedInput class. This function is used only once anyway.
Let's inline this code in the call site.

majetideepak reviewed (commented):
@marin-ma probably `enqueueRowGroup` should always take `input_` and return a pointer to BufferedInput.
Inside `enqueueRowGroup`, we should check if the rowgroup range is already buffered,
if it is, we return `input_`, if not we create a clone of `input_` enqueue, load, and return that. What do you think?

marin-ma commented:
> @marin-ma probably `enqueueRowGroup` should always take `input_` and return a pointer to BufferedInput. Inside `enqueueRowGroup`, we should check if the rowgroup range is already buffered, if it is, we return `input_`, if not we create a clone of `input_` enqueue, load, and return that. What do you think?

Thanks for the review. I agree with your points. This can hide the check from low-level APIs and not breaking the original design. I'll see how to improve the implementation.

marin-ma commented:
@majetideepak Updated. PTAL. Thanks!

majetideepak commented on velox/dwio/parquet/reader/ParquetData.h:
```diff
@@ -172,6 +172,24 @@ class ParquetData : public dwio::common::FormatData {
   }
 
+  // Returns the <offset, length> of the row group.
+  std::pair<int64_t, int64_t> getRowGroupRegion(uint32_t index) const {
```
move the implementation to the cpp file.

majetideepak commented on velox/dwio/parquet/reader/StructColumnReader.h:
```diff
@@ -65,6 +69,10 @@ class StructColumnReader : public dwio::common::SelectiveStructColumnReader {
  private:
   dwio::common::SelectiveColumnReader* findBestLeaf();
 
+  void enqueueRowGroup0(uint32_t index, dwio::common::BufferedInput& input);
```
let's use `enqueueRowGroup` and `loadRowGroup`.

majetideepak reviewed (commented):
@marin-ma Looks much cleaner now. Thanks!
Some comments.

majetideepak commented on velox/dwio/parquet/reader/StructColumnReader.cpp:
```diff
@@ -92,12 +92,36 @@ void StructColumnReader::read(
+  // Check whether the entire row group has already loaded. If true,
+  // reuse the input and data will be read from buffer.
+  // Or else create a new input.
+  auto newInput = isRowGroupBuffered(index, *input) ? input : input->clone();
```
Can simplify as
```
 if(isRowGroupBuffered(index, *input)) {
   return input;
 }
 auto newInput = input->clone();
 enqueueRowGroup(index, *newInput);
 newInput->load(dwio::common::LogType::STRIPE);
 return newInput;
```

marin-ma replied:
Still need to call `enqueueRowGroup` if row group is buffered. It assgins stream_ of the column chunk to the buffered data.
https://github.com/facebookincubator/velox/blob/78c6d45c4ddcb97fcbac30d76fcafaf040b39556/velox/dwio/parquet/reader/ParquetData.cpp#L109

majetideepak reviewed (approved):
@marin-ma thanks!

majetideepak commented:
@Yuhta can you take a look?
